### PR TITLE
Feature: Support for existing boot volume ID

### PR DESCRIPTION
**Pull request by deepaksibm**
Create a custom image of a VSI that is created from existing boot volume

### CODE REVIEW

File: developer/examples/build.vpc.centos.pkr.hcl [hunk 1]
```diff
@@ -46,7 +46,8 @@ source "ibmcloud-vpc" "centos" {
   security_group_id = var.SECURITY_GROUP_ID
 
   vsi_base_image_name = "ibm-centos-7-9-minimal-amd64-5"
-
+    
+  //vsi_boot_volume_id = ""
```
Comment (BShrivastav) [inline developer/examples/build.vpc.centos.pkr.hcl:50]: Would we plan to provide sample use case ?

File: builder/ibmcloud/vpc/config.go [hunk 2]
```diff
@@ -31,6 +31,7 @@ type Config struct {
 	ResourceGroupID           string `mapstructure:"resource_group_id"`
 	SecurityGroupID           string `mapstructure:"security_group_id"`
 	VSIBaseImageID            string `mapstructure:"vsi_base_image_id"`
+	VSIBootVolumeID           string `mapstructure:"vsi_boot_volume_id"`
```
Comment (BShrivastav) [inline builder/ibmcloud/vpc/config.go:34]: Suggestion to align it alphabetically.

**Review by BShrivastav — CHANGES_REQUESTED** [review]

File: builder/ibmcloud/vpc/config.go [hunk 3]
```diff
@@ -84,20 +85,70 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
...
-			errs = packer.MultiErrorAppend(errs, errors.New("one of (vsi_base_image_id or vsi_base_image_name) or (catalog_offering_crn or catalog_offering_version_crn) is required"))
-		} else if c.CatalogOfferingCRN != "" && c.CatalogOfferingVersionCRN != "" {
-			errs = packer.MultiErrorAppend(errs, errors.New("only one of (catalog_offering_crn or catalog_offering_version_crn) is required"))
-		}
-	} else if c.CatalogOfferingCRN == "" && c.CatalogOfferingVersionCRN == "" {
-		if c.VSIBaseImageID != "" && c.VSIBaseImageName != "" {
-			errs = packer.MultiErrorAppend(errs, errors.New("only one of (vsi_base_image_id or vsi_base_image_name) is required"))
-		}
-	} else {
+	// if c.VSIBaseImageID == "" && c.VSIBaseImageName == "" {
```
[the remark this answers is not in the document]
Comment (deepaksibm) [reply builder/ibmcloud/vpc/config.go]: Removed commented code

File: builder/ibmcloud/vpc/config.go [hunk 4]
```diff
@@ -84,20 +85,70 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
...
+		errs = packer.MultiErrorAppend(errs, errors.New("only one of (vsi_base_image_id or vsi_base_image_name) or vsi_boot_volume_id is required"))
+	} else if c.CatalogOfferingCRN != "" && c.CatalogOfferingVersionCRN != "" {
+		errs = packer.MultiErrorAppend(errs, errors.New("only one of (catalog_offering_crn or catalog_offering_version_crn) is required"))
+	} else if c.VSIBaseImageID != "" && c.VSIBaseImageName != "" {
+		errs = packer.MultiErrorAppend(errs, errors.New("only one of (vsi_base_image_id or vsi_base_image_name) is required"))
+	} else if c.CatalogOfferingCRN == "" && c.CatalogOfferingVersionCRN == "" && c.VSIBaseImageID == "" && c.VSIBaseImageName == "" && c.VSIBootVolumeID == "" {
+		errs = packer.MultiErrorAppend(errs, errors.New("one of (vsi_base_image_id or vsi_base_image_name) or (catalog_offering_crn or catalog_offering_version_crn) or vsi_boot_volume_id  is required"))
 	}
 
+	// //Method 3
```
[the remark this answers is not in the document]
Comment (deepaksibm) [reply builder/ibmcloud/vpc/config.go]: Removed the commented code

File: developer/examples/build.vpc.centos.pkr.hcl (hunk 1 above)
Comment (deepaksibm) [reply developer/examples/build.vpc.centos.pkr.hcl:50]: added a new packer template file with this use case. developer/examples/build.vpc.centos.boot.volume.pkr.hcl

File: builder/ibmcloud/vpc/config.go (hunk 2 above)
Comment (deepaksibm) [reply builder/ibmcloud/vpc/config.go]: Moved VSIBootVolumeID below VSIBaseImageName

File: builder/ibmcloud/vpc/config.go [hunk 5]
```diff
@@ -84,20 +85,70 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
...
+	// 	}
+	// } else {
+	// 	if c.VSIBootVolumeID != "" {
+	// 		errs = packer.MultiErrorAppend(errs, errors.New("only one of (vsi_base_image_id or vsi_base_image_name) or (catalog_offering_crn or catalog_offering_version_crn) or vsi_boot_volume_id is required"))
+	// 	} else {
+	// 		errs = packer.MultiErrorAppend(errs, errors.New("only one of (vsi_base_image_id or vsi_base_image_name) or (catalog_offering_crn or catalog_offering_version_crn) is required"))
+	// 	}
+	// }
+
+	//Method 2
```
[the remark this answers is not in the document]
Comment (deepaksibm) [reply builder/ibmcloud/vpc/config.go]: This includes validation for the five one ofs. anything else should be added to the validation if missing ?

File: developer/variables.pkrvars.boot.volume.hcl [hunk 6]
```diff
@@ -0,0 +1,6 @@
+SUBNET_ID="<subnet_id>"
+REGION="us-south"
```
Comment (BShrivastav) [inline developer/variables.pkrvars.boot.volume.hcl:2]: Remove the unwanted values

File: developer/variables.pkrvars.boot.volume.hcl [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+SECURITY_GROUP_ID=""
+RESOURCE_GROUP_ID=""
+IBM_API_KEY="<ibm_api_key>"
+BOOT_VOLUME_ID="boot_volume_id"
```
Comment (BShrivastav) [inline developer/variables.pkrvars.boot.volume.hcl:6]:
```suggestion
BOOT_VOLUME_ID="<boot_volume_id>"
```

**Review by BShrivastav — CHANGES_REQUESTED** [review]
Added some suggestions.

**Review by ujjwal-ibm — CHANGES_REQUESTED** [review]

File: developer/Makefile [hunk 8]
```diff
@@ -31,6 +31,10 @@ run-catalog:
 run-centos:
 	cd ..; packer validate -var-file="developer/variables.pkrvars.hcl" developer/examples/build.vpc.centos.pkr.hcl
 	cd ..; packer build -var-file="developer/variables.pkrvars.hcl" developer/examples/build.vpc.centos.pkr.hcl
+
+run-centos-boot-volume:
```
[the remark this answers is not in the document]
Comment (deepaksibm) [reply developer/Makefile]: modified the name

File: developer/Makefile [hunk 9]
```diff
@@ -31,6 +31,10 @@ run-catalog:
...
+	cd ..; packer validate -var-file="developer/variables.pkrvars.boot.volume.hcl" developer/examples/build.vpc.centos.boot.volume.pkr.hcl
+	cd ..; packer build -var-file="developer/variables.pkrvars.boot.volume.hcl" developer/examples/build.vpc.centos.boot.volume.pkr.hcl
```
[the remark this answers is not in the document]
Comment (deepaksibm) [reply developer/Makefile]: added to validate-all

File: developer/variables.boot.volume.pkrvars.hcl [hunk 10]
```diff
@@ -0,0 +1,6 @@
+SUBNET_ID="0727-9447a2c3-efb8-44d2-a180-115531e0b0d3"
```
Comment (BShrivastav) [inline developer/variables.boot.volume.pkrvars.hcl:1]: Can we remove references int file and follow standard ones.

**Review by BShrivastav — CHANGES_REQUESTED** [review]
Cleanup change for the file - developer/variables.boot.volume.pkrvars.hcl and we should be good.

File: developer/variables.boot.volume.pkrvars.hcl [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+REGION="us-south"
+SECURITY_GROUP_ID=""
+RESOURCE_GROUP_ID=""
+IBM_API_KEY="UUXUGZf3BSWd-kVJHLqBJVfirD2xPphT8ecK8nTbj-Jg"
```
Comment (syyang-in-cloud) [inline developer/variables.boot.volume.pkrvars.hcl:5]: IBM_API_KEY is exposed here. If this is a real API KEY, then you should revoke the API KEY. 
For the future, run `detect-secrets` as `pre-commit` hook.

**Review by syyang-in-cloud — CHANGES_REQUESTED** [review]
Please revoke the exposed API KEY.

File: developer/variables.boot.volume.pkrvars.hcl (hunk 11 above)
Comment (deepaksibm) [reply developer/variables.boot.volume.pkrvars.hcl]: revoked this change

File: developer/variables.boot.volume.pkrvars.hcl (hunk 10 above)
Comment (deepaksibm) [reply developer/variables.boot.volume.pkrvars.hcl]: revoked this change

**Comment by deepaksibm** [comment]
> 

revoked

**Comment by deepaksibm** [comment]
> Please revoke the exposed API KEY.

revoked

**Review by astha-jain — APPROVED** [review]
Looks good

**Review by BShrivastav — APPROVED** [review]
Looks good.

File: README.md [hunk 12]
```diff
@@ -194,6 +194,8 @@ vsi_base_image_name | string | The base image name used to created the VSI. Use
 catalog_offering_crn | string | The [catalog](https://cloud.ibm.com/docs/account?topic=account-restrict-by-user) offering version to use when provisioning this virtual server instance. The specified offering version may be in a different account in the same enterprise, subject to IAM policies. Identifies a [catalog](https://cloud.ibm.com/docs/account?topic=account-restrict-by-user) offering by a unique property. Optional.
 | OR |
 catalog_offering_version_crn | string | The [catalog](https://cloud.ibm.com/docs/account?topic=account-restrict-by-user) offering version to use when provisioning this virtual server instance. The specified offering version may be in a different account in the same enterprise, subject to IAM policies. Identifies a version of a [catalog](https://cloud.ibm.com/docs/account?topic=account-restrict-by-user) offering by a unique property. Optional.
+| OR |
+vsi_boot_volume_id | string | The ID of the boot volume to be used to create the VSI. Use `ibmcloud is volumes` for available options. It should be a boot volume which has operating system details in it.
```
Comment (astha-jain) [inline README.md:198]: The ID of the an existing boot volume. 
~~It should be a boot volume which has operating system details in it.~~

Comment (deepaksibm) [reply README.md]: updated

File: developer/examples/build.vpc.boot.volume.pkr.hcl [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  api_key           = var.IBM_API_KEY
+  region            = var.REGION
+  subnet_id         = var.SUBNET_ID
+  resource_group_id = var.RESOURCE_GROUP_ID
+  security_group_id = var.SECURITY_GROUP_ID
+    
+  vsi_boot_volume_id = var.BOOT_VOLUME_ID
+  vsi_profile        = "bx2-2x8"
+  vsi_interface      = "public"
+  vsi_user_data_file = ""
```
Comment (astha-jain) [inline developer/examples/build.vpc.boot.volume.pkr.hcl:56]: What will be the impact of `vsi_user_data_file = ""`?

**Review by astha-jain — APPROVED** [review]

**Review by astha-jain — APPROVED** [review]

**Pull request merged**